src/compiler/checker.ts

```diff
@@ -3062,6 +3064,26 @@ namespace ts {
             setObjectTypeMembers(type, emptySymbols, callSignatures, constructSignatures, stringIndexType, numberIndexType);
         }
 
+        function intersectTypes(type1: Type, type2: Type): Type {
+            return !type1 ? type2 : !type2 ? type1 : getIntersectionType([type1, type2]);

```

Comment:
I actually find the way @ahejlsberg wrote it clearer than the `&&` style. But I'm also okay with the longhand you suggested.